### PR TITLE
Update rustls to 0.23 and ring to 0.17

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: quinn-proto/Cargo.toml [hunk 1]
```diff
@@ -17,8 +17,9 @@ all-features = true
 maintenance = { status = "experimental" }
 
 [features]
-default = ["tls-rustls", "log"]
+default = ["tls-rustls", "rustls-ring", "log"]
 tls-rustls = ["rustls", "ring"]
+rustls-ring = ["rustls/ring"]
```
Comment (Ralith) [inline quinn-proto/Cargo.toml:22]: Document the use case for this? It allows downstream code using some other rustls backend to avoid a ring dependency, right?

File: quinn-proto/src/config.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 use std::{fmt, num::TryFromIntError, sync::Arc, time::Duration};
 
+use rustls_pki_types::{CertificateDer, PrivateKeyDer};
```
Comment (Ralith) [inline quinn-proto/src/config.rs:3]: `rustls-pki-types` is optional, so should this be cfg-gated?

File: quinn-proto/src/config.rs [hunk 3]
```diff
@@ -829,8 +830,8 @@ impl ServerConfig {
     ///
     /// Uses a randomized handshake token key.
     pub fn with_single_cert(
-        cert_chain: Vec<rustls::Certificate>,
-        key: rustls::PrivateKey,
+        cert_chain: Vec<CertificateDer<'static>>,
+        key: PrivateKeyDer<'static>,
```
Comment (Ralith) [inline quinn-proto/src/config.rs:834]: How do we know `rustls-pki-types` is enabled?

File: quinn-proto/src/crypto/rustls.rs [hunk 4]
```diff
@@ -386,10 +400,6 @@ impl crypto::PacketKey for PacketKey {
 /// satisfies this requirement.
 pub(crate) fn client_config(roots: rustls::RootCertStore) -> rustls::ClientConfig {
     let mut cfg = rustls::ClientConfig::builder()
-        .with_safe_default_cipher_suites()
-        .with_safe_default_kx_groups()
-        .with_protocol_versions(&[&rustls::version::TLS13])
-        .unwrap()
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:392]: Why the change here? In particular, is TLS1.3 only the default upstream now?

File: quinn-proto/src/crypto/rustls.rs [hunk 5]
```diff
@@ -204,7 +206,7 @@ const RETRY_INTEGRITY_NONCE_V1: [u8; 12] = [
     0x46, 0x15, 0x99, 0xd3, 0x5d, 0x63, 0x2b, 0xf2, 0x23, 0x98, 0x25, 0xbb,
 ];
 
-impl crypto::HeaderKey for HeaderProtectionKey {
+impl crypto::HeaderKey for Box<dyn HeaderProtectionKey> {
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:209]: Is there a way we could avoid double-boxing here?

File: quinn-proto/src/config.rs (hunk 3 above)
Comment (djc) [reply quinn-proto/src/config.rs:834]: Because `feature = "rustls"` will now also enable `feature = "rustls-pki-types"`.

File: quinn-proto/src/crypto/rustls.rs (hunk 5 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:209]: I haven't figured out a way. I'm afraid it might be pretty hard and there are still other open issues.

File: quinn-proto/src/crypto/rustls.rs (hunk 4 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:392]: IIRC rustls no longer needs a TLS 1.3 only config for QUIC to work, so there doesn't seem to be much of an advantage to doing this?

File: quinn-proto/src/crypto/rustls.rs (hunk 5 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:209]: Yeah, I think it's impossible, short of reimplementing a good chunk of rustls's QUIC logic to undo rustls's backend abstraction. Not a huge deal.

File: quinn-proto/src/crypto/rustls.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:392]: Enabling TLS 1.2, which is strictly illegal to use in QUIC, feels pretty weird.

File: quinn-proto/Cargo.toml [hunk 6]
```diff
@@ -17,8 +17,9 @@ all-features = true
 maintenance = { status = "experimental" }
 
 [features]
-default = ["tls-rustls", "log"]
-tls-rustls = ["rustls", "ring"]
+default = ["tls-rustls", "rustls-ring", "log"]
+tls-rustls = ["rustls-pki-types"]
+rustls-ring = ["rustls/ring", "ring"] # Allow rustls to be used with other crypto providers
```
Comment (Ralith) [inline quinn-proto/Cargo.toml:22]: This wording makes it seem like you should *enable* the feature if you want that.
```suggestion
# Disable to allow rustls to be used with other crypto providers enabled elsewhere
rustls-ring = ["rustls/ring", "ring"]
```

File: quinn-proto/src/crypto/rustls.rs (hunk 4 above)
Comment (cpu) [reply quinn-proto/src/crypto/rustls.rs:392]: It should be possible to use [`rustls::ClientConfig::builder_with_protocol_versions`](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.with_protocol_versions) here to get the default crypto provider but explicitly specify TLS 1.3 only.

**Comment by getong** [comment]
I think you might also change the code in directory `quinn/examples`, like server.rs and client.rs.

rustls::Certificate has been replaced with CertificateDer from the new [rustls-pki-types](https://docs.rs/rustls-pki-types/latest/rustls_pki_types/) crate. 
Likewise, rustls::PrivateKey has been replaced with rustls_pki_types::PrivateKeyDer.

The code in examples does not compile.

**Comment by djc** [comment]
@getong there is a reason the PR is in draft state.

**Comment by PikuZheng** [comment]
In order to make ring support some architectures, I hope to upgrade ring to 0.17 as soon as possible

**Comment by cpu** [comment]
> In order to make ring support some architectures, I hope to upgrade ring to 0.17 as soon as possible

@PikuZheng I believe *ring* 0.17 is _already_ available with Quinn. The existing Quinn 0.10 release uses Rustls 0.21, which has upgraded to *ring* 0.17 as of  [v0.21.8](https://github.com/rustls/rustls/releases/tag/v%2F0.21.8. You don't need to wait for Quinn to have Rustls 0.22 support for this.

**Comment by PikuZheng** [comment]
> > In order to make ring support some architectures, I hope to upgrade ring to 0.17 as soon as possible
> 
> @PikuZheng I believe _ring_ 0.17 is _already_ available with Quinn. The existing Quinn 0.10 release uses Rustls 0.21, which has upgraded to _ring_ 0.17 as of [v0.21.8](https://github.com/rustls/rustls/releases/tag/v%2F0.21.8). You don't need to wait for Quinn to have Rustls 0.22 support for this.

```
├── quinn v0.10.2
│   ├── bytes v1.5.0
│   ├── pin-project-lite v0.2.13
│   ├── quinn-proto v0.10.6
│   │   ├── bytes v1.5.0
│   │   ├── rand v0.8.5
│   │   │   ├── libc v0.2.151
│   │   │   ├── rand_chacha v0.3.1
│   │   │   │   ├── ppv-lite86 v0.2.17
│   │   │   │   └── rand_core v0.6.4
│   │   │   │       └── getrandom v0.2.11 (*)
│   │   │   └── rand_core v0.6.4 (*)
│   │   ├── ring v0.16.20
│   │   │   ├── libc v0.2.151
│   │   │   ├── once_cell v1.19.0
│   │   │   ├── spin v0.5.2
│   │   │   └── untrusted v0.7.1
│   │   ├── rustc-hash v1.1.0
│   │   ├── rustls v0.21.10 (*)
│   │   ├── slab v0.4.9 (*)
│   │   ├── thiserror v1.0.52 (*)
│   │   ├── tinyvec v1.6.0
│   │   │   └── tinyvec_macros v0.1.1
│   │   └── tracing v0.1.40 (*)
```
anything i'm missing?

**Comment by cpu** [comment]
> anything i'm missing?

No, this was my mistake. I forgot Quinn also has *ring* in its public API and didn't take the breaking change to update to 0.17 in the 0.10 release like Rustls did.

Sorry for the confusion. There is a change required in this repo to remove ring 0.16 from the requirements.

**Comment by djc** [comment]
We can now compile quinn-proto -- but there's still a bunch of test failures:

```
failures:
    tests::handshake_anti_deadlock_probe
    tests::large_initial
    tests::reject_missing_client_cert
    tests::server_can_send_3_inital_packets
```

If someone wants to lend a hand diagnosing/fixing these, that would be great. Interestingly, some of these seem to be triggering panics in the rustls deframer:

```
thread 'tests::server_can_send_3_inital_packets' panicked at /Users/djc/.cargo/git/checkouts/rustls-29bde2ca1adbde2d/aef485d/rustls/src/msgs/deframer.rs:490:8:
range end index 4468 out of range for slice of length 4096
```

File: quinn-proto/src/crypto.rs [hunk 7]
```diff
@@ -209,16 +210,13 @@ pub trait AeadKey {
...
-    }
+/// Crypto layer errors
+#[derive(Clone, Copy, Debug, Eq, Error, PartialEq)]
+pub enum CryptoError {
+    /// Error from the cryptographic session layer
+    #[error("crypto error")]
+    Other,
+    /// Specified QUIC version is not supported
+    #[error("unsupported QUIC protocol version")]
+    UnsupportedVersion,
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs:216]: Why combine the errors? I don't see any new error cases in `initial_keys`, and it's nice to express that guarantee.

File: Cargo.toml [hunk 8]
```diff
@@ -8,3 +8,6 @@ debug = true
 
 [profile.release]
 debug = true
+
+[patch."crates-io"]
+rustls = { git = "https://github.com/rustls/rustls", rev = "aef485d54bc1c15ffb98db5714a2e32df814c9e7" }
```
Comment (Ralith) [inline Cargo.toml:13]: Is this still needed?

File: quinn-proto/src/crypto/rustls.rs [hunk 9]
```diff
@@ -260,13 +263,14 @@ impl crypto::ClientConfig for rustls::ClientConfig {
                 rustls::quic::ClientConnection::new(
                     self,
                     version,
-                    server_name
-                        .try_into()
-                        .map_err(|_| ConnectError::InvalidDnsName(server_name.into()))?,
+                    ServerName::try_from(server_name)
+                        .map_err(|_| ConnectError::InvalidDnsName(server_name.into()))?
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:267]: Should we rename this error variant now that `server_name` can be an IP?

File: quinn-proto/src/crypto/rustls.rs [hunk 10]
```diff
@@ -278,19 +282,23 @@ impl crypto::ServerConfig for rustls::ServerConfig {
...
                 rustls::quic::ServerConnection::new(self, version, to_vec(params)).unwrap(),
             ),
+            suite,
         })
     }
 
     fn initial_keys(&self, version: u32, dst_cid: &ConnectionId) -> Result<Keys, CryptoError> {
         let version = interpret_version(version)?;
-        Ok(initial_keys(version, dst_cid, Side::Server))
+        // We validate that this works on configuration, so we can `unwrap()` here.
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:299]: I'm not sure what "on configuration" means here. Can we cite the responsible code more explicitly?

**Review by Ralith — COMMENTED** [review]
Interesting failures; seems like we must be mangling rustls's data somehow, or the upstream API broke. Will try to poke around, time permitting.

File: Cargo.toml (hunk 8 above)
Comment (Ralith) [reply Cargo.toml:13]: Ah, I see there's a PR.

**Comment by Ralith** [comment]
This patch to rustls fixes 3/4 failing tests:

```diff
diff --git a/rustls/src/msgs/deframer.rs b/rustls/src/msgs/deframer.rs
index 6da2d9d4..5d9d1fdd 100644
--- a/rustls/src/msgs/deframer.rs
+++ b/rustls/src/msgs/deframer.rs
@@ -253,7 +253,7 @@ impl MessageDeframer {
                 // We're joining a handshake message to the previous one here.
                 // Write it into the buffer and update the metadata.
 
-                DeframerBuffer::<QUIC>::copy(buffer, payload, meta.payload.end);
+                DeframerBuffer::<QUIC>::copy(buffer, payload, 0);
                 meta.message.end = end;
                 meta.payload.end += payload.len();
```

`DeframerBuffer::<true>::copy` automatically copies into the end of the payload, so passing `meta.payload.end` as the third argument double-counts the offset. This branch is probably only covered by tests which use multiple calls to `read_hs` to pass in a single handshake message.

**Comment by Ralith** [comment]
This fixes the remaining test:
```diff
diff --git a/quinn-proto/src/tests/mod.rs b/quinn-proto/src/tests/mod.rs
index f809ba1c..1b3b28de 100644
--- a/quinn-proto/src/tests/mod.rs
+++ b/quinn-proto/src/tests/mod.rs
@@ -379,9 +379,15 @@ fn reject_missing_client_cert() {
     let key = PrivatePkcs8KeyDer::from(CERTIFICATE.serialize_private_key_der());
     let cert = CertificateDer::from(util::CERTIFICATE.serialize_der().unwrap());
 
+    let mut store = RootCertStore::empty();
+    // `WebPkiClientVerifier` requires a nonempty store, so we stick our own certificate into it
+    // because it's convenient.
+    store
+        .add(CERTIFICATE.serialize_der().unwrap().into())
+        .unwrap();
     let config = rustls::ServerConfig::builder()
         .with_client_cert_verifier(
-            WebPkiClientVerifier::builder(Arc::new(RootCertStore::empty()))
+            WebPkiClientVerifier::builder(Arc::new(store))
                 .build()
                 .unwrap(),
         )
```

File: .github/workflows/rust.yml [hunk 11]
```diff
@@ -2,7 +2,7 @@ name: CI
 
 on:
   push:
-    branches: ['main', '0.8.x']
+    branches: ['main', '0.8.x', 'cpu-*']
```
Comment (cpu) [inline .github/workflows/rust.yml:5]:
```suggestion
    branches: ['main', '0.8.x']
```

File: Cargo.toml [hunk 12]
```diff
@@ -8,3 +8,7 @@ debug = true
 
 [profile.release]
 debug = true
+
+[patch."crates-io"]
+rustls = { git = "https://github.com/cpu/rustls", rev = "9d547f70ba60fe3f3b5e68f41f14b9354c545ff1" }
```
Comment (cpu) [inline Cargo.toml:13]:
```suggestion
rustls = { git = "https://github.com/rustls/rustls", rev = "d5842f461e3049d4e3857a60a083f65158a303be" }
```

**Comment by mokeyish** [comment]
Looking forward to merging this PR soon.

**Comment by djc** [comment]
I think this is blocked on the rustls-platform-verifier upgrade to rustls 0.23 for now.

File: quinn-proto/src/crypto/rustls.rs [hunk 13]
```diff
@@ -59,9 +61,9 @@ impl crypto::Session for TlsSession {
     }
 
     fn peer_identity(&self) -> Option<Box<dyn Any>> {
-        self.inner
-            .peer_certificates()
-            .map(|v| -> Box<dyn Any> { Box::new(v.to_vec()) })
+        self.inner.peer_certificates().map(|v| -> Box<dyn Any> {
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:66]: rustls gives us `CertificateDer<'_>` elements here which trivially convert from DER represented as `&[u8]` or `Vec<u8>`. How do you feel about just letting that leak into the public API? (Particularly given that rustls-pki-types is also intended as a long-term stable API.)

File: quinn-proto/src/crypto.rs (hunk 7 above)
Comment (djc) [reply quinn-proto/src/crypto.rs]: Yeah, I think there was a reason when I was experimenting with getting this working against 0.22, but it doesn't seem to be needed anymore.

File: quinn-proto/src/crypto/rustls.rs (hunk 9 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs]: Done.

File: quinn-proto/src/crypto/rustls.rs [hunk 14]
```diff
@@ -256,6 +256,7 @@ impl crypto::ClientConfig for rustls::ClientConfig {
         params: &TransportParameters,
     ) -> Result<Box<dyn crypto::Session>, ConnectError> {
         let version = interpret_version(version)?;
+        // TODO: why is this unwrap safe?
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:259]: Is this safe, and if so, why?

File: quinn-proto/src/crypto/rustls.rs [hunk 15]
```diff
@@ -283,7 +284,9 @@ impl crypto::ServerConfig for rustls::ServerConfig {
         version: u32,
         params: &TransportParameters,
     ) -> Box<dyn crypto::Session> {
+        // TODO: why is this unwrap safe?
         let version = interpret_version(version).unwrap();
+        // TODO: why is this unwrap safe?
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:289]: Are these safe, and if so, why?

File: quinn-proto/src/crypto/rustls.rs [hunk 16]
```diff
@@ -302,11 +305,13 @@ impl crypto::ServerConfig for rustls::ServerConfig {
         dst_cid: &ConnectionId,
     ) -> Result<Keys, UnsupportedVersion> {
         let version = interpret_version(version)?;
+        // TODO: why is this unwrap safe?
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:308]: Is this safe, and if so, why?

**Review by cpu — APPROVED** [review]
I looked through the final diff after all the downstream work shook out and it looks good to me (_with the caveat that I contributed some of it, and don't have much background on Quinn or QUIC generally!_)

File: quinn-proto/src/crypto/rustls.rs (hunk 13 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:66]: rustls (when used) is a public dep no matter what, so I'm not opposed to exposing stuff from it. We're currently exposing a `rustls::Certificate` here anyway, after all.

Let's be sure to update the comment in `quinn::Connection::peer_identity` to reflect the actual type, if that's changing. Maybe also fill in the `_` in the `collect` call below for clarity? It's hard to tell what the type is after this PR.

File: quinn-proto/src/crypto/rustls.rs (hunk 14 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:259]: Seems to me that this could panic if someone sets a funny provider. Maybe we should return an error here, similar to how `interpret_version` does above?

File: quinn-proto/src/crypto/rustls.rs (hunk 15 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:289]: `interpret_version` is safe per the docs on the declaration of `ServerConfig::start_session`:

> Never called if `initial_keys` rejected `version`.

The unwrap on `suite_from_provider` is not. Maybe we should validate it similarly?

File: quinn-proto/src/crypto/rustls.rs (hunk 16 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:308]: I don't see why it would be. I think we should handle it similarly to `initial_version` above.

File: quinn-proto/src/crypto/rustls.rs [hunk 17]
```diff
@@ -302,11 +305,13 @@ impl crypto::ServerConfig for rustls::ServerConfig {
...
         let suite = suite_from_provider(self.crypto_provider()).unwrap();
         Ok(initial_keys(version, dst_cid, Side::Server, &suite))
     }
 
     fn retry_tag(&self, version: u32, orig_dst_cid: &ConnectionId, packet: &[u8]) -> [u8; 16] {
+        // TODO: why is this unwrap safe?
```
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:314]: Yes. Just like with `start_session`, the docs say:

> Never called if `initial_keys` rejected `version`.

**Comment by gabrik** [comment]
Hi @djc , should I do a fork and then open a PR targeting this branch or should I ask access?

**Comment by djc** [comment]
Probably just a fork + PR targeting this branch?

**Comment by gabrik** [comment]
Done PR here: https://github.com/quinn-rs/quinn/pull/1808

**Comment by gabrik** [comment]
PR: https://github.com/quinn-rs/quinn/pull/1809

File: quinn-proto/src/crypto/rustls.rs (hunk 15 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs]: We now have derived `QuicClientConfig` and `QuicServerConfig` types that can be fallibly constructed from an existing rustls `ClientConfig` or `ServerConfig`, respectively.

File: quinn-proto/src/crypto/rustls.rs (hunk 13 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:68]: Decided to leave this for now, since for example a Noise `crypto::Session` still wouldn't want a `CertificateDer<'_>` to represent its identitifes. Did document the type for the rustls implementation.

File: quinn-proto/src/tests/mod.rs [hunk 18]
```diff
@@ -603,9 +603,13 @@ fn zero_rtt_rejection() {
     pair.client.connections.clear();
     pair.server.connections.clear();
 
+    // Need to modify ALPN protocols in the existing client crypto config to enable resumption
```
Comment (Ralith) [inline quinn-proto/src/tests/mod.rs:608]: nit: IIRC the only thing that we really need to preserve here is the session cache, but this comment suggests otherwise.

File: quinn-proto/src/crypto/rustls.rs [hunk 19]
```diff
@@ -357,24 +357,81 @@ impl std::fmt::Display for NoInitialCipherSuite {
 
 impl std::error::Error for NoInitialCipherSuite {}
 
-impl crypto::ServerConfig for rustls::ServerConfig {
+/// A QUIC-compatible TLS server configuration
+pub struct QuicServerConfig {
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:361]: Document `TryFrom`

File: quinn-proto/src/crypto/rustls.rs [hunk 20]
```diff
@@ -357,24 +357,81 @@ impl std::fmt::Display for NoInitialCipherSuite {
...
+        .with_single_cert(cert_chain, key)
+        .unwrap();
+
+        inner.max_early_data_size = u32::MAX;
+        inner
+    }
+}
+
+impl TryFrom<rustls::ServerConfig> for QuicServerConfig {
+    type Error = NoInitialCipherSuite;
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:404]: Consider validating TLS 1.3 support

File: quinn-proto/src/crypto/rustls.rs [hunk 21]
```diff
@@ -357,24 +357,81 @@ impl std::fmt::Display for NoInitialCipherSuite {
...
+    /// `u32::MAX`. Advanced users can use any [`rustls::ServerConfig`] that satisfies these
+    /// requirements.
+    pub(crate) fn inner(
+        cert_chain: Vec<CertificateDer<'static>>,
+        key: PrivateKeyDer<'static>,
+    ) -> rustls::ServerConfig {
+        let mut inner = rustls::ServerConfig::builder_with_provider(
+            rustls::crypto::ring::default_provider().into(),
+        )
+        .with_safe_default_protocol_versions()
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:392]: This should explicitly enable only TLS 1.3.

**Review by Ralith — CHANGES_REQUESTED** [review]

**Comment by Ralith** [comment]
Made a deep pass; looks like only some minor cosmetic issues left.

File: quinn-proto/src/crypto/rustls.rs (hunk 20 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:404]: I don't think the rustls configs currently support interrogating version support.

File: quinn-proto/src/crypto/rustls.rs (hunk 21 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:392]: Fixed.

File: quinn-proto/src/crypto/rustls.rs (hunk 19 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:361]: Done.

File: quinn-proto/src/crypto/rustls.rs [hunk 22]
```diff
@@ -275,11 +319,44 @@ impl crypto::ClientConfig for rustls::ClientConfig {
...
+            suite: Suite {
+                suite: self.initial.suite,
+                quic: self.initial.quic,
+            },
         }))
     }
 }
 
+impl TryFrom<rustls::ClientConfig> for QuicClientConfig {
+    type Error = NoInitialCipherSuite;
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:331]: rustls doesn't currently expose API for this.

File: quinn-proto/src/crypto/rustls.rs [hunk 23]
```diff
@@ -250,23 +250,67 @@ pub struct HandshakeData {
     pub server_name: Option<String>,
 }
 
-impl crypto::ClientConfig for rustls::ClientConfig {
+/// A QUIC-compatible TLS client configuration
+pub struct QuicClientConfig {
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:254]: Added.

File: quinn-proto/src/crypto/rustls.rs [hunk 24]
```diff
@@ -383,16 +384,29 @@ impl crypto::PacketKey for PacketKey {
...
+            .with_custom_certificate_verifier(Arc::new(rustls_platform_verifier::Verifier::new()))
+            .with_no_client_auth(),
+        ServerVerifier::Roots(roots) => builder.with_root_certificates(roots).with_no_client_auth(),
+    };
+
+    config.enable_early_data = true;
+    config
+}
+
+pub(crate) enum ServerVerifier {
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs]: Replaced this by passing an `Arc<dyn ServerCertVerifier>`, there wasn't really a good reason.

File: quinn-proto/src/crypto/rustls.rs [hunk 25]
```diff
@@ -61,10 +63,11 @@ impl crypto::Session for TlsSession {
         }))
     }
 
+    /// For the rustls `TlsSession`, the type of the `Any` is `rustls::pki_types::CertificateDer`
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:66]: Fixed.

File: quinn-proto/src/crypto/rustls.rs [hunk 26]
```diff
@@ -61,10 +63,11 @@ impl crypto::Session for TlsSession {
...
     fn peer_identity(&self) -> Option<Box<dyn Any>> {
-        self.inner
-            .peer_certificates()
-            .map(|v| -> Box<dyn Any> { Box::new(v.to_vec()) })
+        self.inner.peer_certificates().map(|v| -> Box<dyn Any> {
+            Box::new(v.iter().map(|v| v.clone().into_owned()).collect::<Vec<_>>())
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:69]: Fixed.

File: quinn-proto/src/crypto/rustls.rs [hunk 27]
```diff
@@ -415,16 +447,16 @@ pub(crate) enum ServerVerifier {
...
-        .with_safe_default_cipher_suites()
-        .with_safe_default_kx_groups()
-        .with_protocol_versions(&[&rustls::version::TLS13])
-        .unwrap()
-        .with_no_client_auth()
-        .with_single_cert(cert_chain, key)?;
+    let mut cfg = rustls::ServerConfig::builder_with_provider(
+        rustls::crypto::ring::default_provider().into(),
+    )
+    .with_safe_default_protocol_versions()
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs]: Fixed this.

File: quinn-proto/src/tests/mod.rs (hunk 18 above)
Comment (djc) [reply quinn-proto/src/tests/mod.rs]: Extended the comment to more clearly explain the constraints.

**Comment by djc** [comment]
@Ralith I think this is ready for another round -- hoping it will be the last one!

File: quinn-proto/src/crypto/rustls.rs (hunk 20 above)
Comment (Ralith) [reply quinn-proto/src/crypto/rustls.rs:404]: Aw. Would be nice to have, but not a blocker.

File: quinn-proto/src/crypto/rustls.rs [hunk 28]
```diff
@@ -260,47 +311,141 @@ impl crypto::ClientConfig for rustls::ClientConfig {
...
+        f.write_str("no initial cipher suite found")
+    }
+}
+
+impl std::error::Error for NoInitialCipherSuite {}
+
+/// A QUIC-compatible TLS server configuration
+///
+/// Can be constructed via [`ServerConfig::with_single_cert()`][single] or by using the
+/// [`TryFrom`] implementation with a custom [`rustls::ServerConfig`].
```
Comment (Ralith) [inline quinn-proto/src/crypto/rustls.rs:358]: We should document requirements on custom `rustls::ServerConfig`s: that TLS 1.3 be enabled exclusively, and that the early data limit be either `0` if 0-RTT is unwanted or `u32::MAX` otherwise.

**Review by Ralith — APPROVED** [review]

File: quinn-proto/src/crypto/rustls.rs (hunk 28 above)
Comment (djc) [reply quinn-proto/src/crypto/rustls.rs]: Extended the docstring here.

**Comment by djc** [comment]
Release progress is tracked in #1737.
